File: utils/weather-helper.c
#include "weather-helper.h"
#include "icons.h"

const char *weather_code_to_description(LAYER_SHELL_IO_WeatherCode code) {
  switch (code) {
  case ClearSky:
    return "Clear Sky";
  case MainlyClear:
    return "Mainly Clear";
  case PartlyCloudy:
    return "Partly Cloudy";
  case Overcast:
    return "Overcast";
  case FogDepositingRime:
    return "Fog (Depositing Rime)";
  case FogNormal:
    return "Fog (Normal)";
  case DrizzleLight:
    return "Drizzle (Light)";
  case DrizzleModerate:
    return "Drizzle (Moderate)";
  case DrizzleDense:
    return "Drizzle (Dense)";
  case FreezingDrizzleLight:
    return "Freezing Drizzle (Light)";
  case FreezingDrizzleDense:
    return "Freezing Drizzle (Dense)";
  case RainSlight:
    return "Rain (Slight)";
  case RainModerate:
    return "Rain (Moderate)";
  case RainHeavy:
    return "Rain (Heavy)";
  case FreezingRainLight:
    return "Freezing Rain (Light)";
  case FreezingRainHeavy:
    return "Freezing Rain (Heavy)";
  case SnowFallSlight:
    return "Snow Fall (Slight)";
  case SnowFallModerate:
    return "Snow Fall (Moderate)";
  case SnowFallHeavy:
    return "Snow Fall (Heavy)";
  case SnowGrains:
    return "Snow Grains";
  case RainShowersSlight:
    return "Rain Showers (Slight)";
  case RainShowersModerate:
    return "Rain Showers (Moderate)";
  case RainShowersViolent:
    return "Rain Showers (Violent)";
  case SnowShowersSlight:
    return "Snow Showers (Slight)";
  case SnowShowersHeavy:
    return "Snow Showers (Heavy)";
  case Thunderstorm:
    return "Thunderstorm";
  case ThunderstormWithHailSight:
    return "Thunderstorm With Hail (Sight)";
  case ThunderstormWithHailHeavy:
    return "Thunderstorm With Hail (Heavy)";
  case Unknown:
    return "Unknown";
  }

  return NULL;
}

GIcon *weather_code_to_icon(LAYER_SHELL_IO_WeatherCode code) {
  switch (code) {
  case ClearSky:
  case MainlyClear:
    return get_sunny_icon();
  case PartlyCloudy:
  case Overcast:
    return get_partly_cloudy_icon();
  case FogDepositingRime:
  case FogNormal:
    return get_foggy_icon();
  case DrizzleDense:
  case DrizzleLight:
  case DrizzleModerate:
  case FreezingDrizzleLight:
  case FreezingDrizzleDense:
  case RainSlight:
  case RainModerate:
  case RainHeavy:
  case FreezingRainLight:
  case FreezingRainHeavy:
  case RainShowersSlight:
  case RainShowersModerate:
  case RainShowersViolent:
    return get_rainy_icon();
  case SnowFallSlight:
  case SnowFallModerate:
  case SnowFallHeavy:
  case SnowGrains:
  case SnowShowersSlight:
  case SnowShowersHeavy:
    return get_snowy_icon();
  case Thunderstorm:
  case ThunderstormWithHailSight:
  case ThunderstormWithHailHeavy:
    return get_thunderstorm_icon();
  case Unknown:
    return get_question_mark_icon();
  }

  return NULL;
}

File: ui/utils/weather_helper.py
from liblayer_shell_io import WeatherCode
from utils.context import ctx


class WeatherHelper:
    @staticmethod
    def code_to_description(code: WeatherCode):
        match code:
            case WeatherCode.ClearSky:
                return "Clear Sky"
            case WeatherCode.MainlyClear:
                return "Mainly Clear"
            case WeatherCode.PartlyCloudy:
                return "Partly Cloudy"
            case WeatherCode.Overcast:
                return "Overcast"
            case WeatherCode.FogDepositingRime:
                return "Fog (Depositing Rime)"
            case WeatherCode.FogNormal:
                return "Fog (Normal)"
            case WeatherCode.DrizzleLight:
                return "Drizzle (Light)"
            case WeatherCode.DrizzleModerate:
                return "Drizzle (Moderate)"
            case WeatherCode.DrizzleDense:
                return "Drizzle (Dense)"
            case WeatherCode.FreezingDrizzleLight:
                return "Freezing Drizzle (Light)"
            case WeatherCode.FreezingDrizzleDense:
                return "Freezing Drizzle (Dense)"
            case WeatherCode.RainSlight:
                return "Rain (Slight)"
            case WeatherCode.RainModerate:
                return "Rain (Moderate)"
            case WeatherCode.RainHeavy:
                return "Rain (Heavy)"
            case WeatherCode.FreezingRainLight:
                return "Freezing Rain (Light)"
            case WeatherCode.FreezingRainHeavy:
                return "Freezing Rain (Heavy)"
            case WeatherCode.SnowFallSlight:
                return "Snow Fall (Slight)"
            case WeatherCode.SnowFallModerate:
                return "Snow Fall (Moderate)"
            case WeatherCode.SnowFallHeavy:
                return "Snow Fall (Heavy)"
            case WeatherCode.SnowGrains:
                return "Snow Grains"
            case WeatherCode.RainShowersSlight:
                return "Rain Showers (Slight)"
            case WeatherCode.RainShowersModerate:
                return "Rain Showers (Moderate)"
            case WeatherCode.RainShowersViolent:
                return "Rain Showers (Violent)"
            case WeatherCode.SnowShowersSlight:
                return "Snow Showers (Slight)"
            case WeatherCode.SnowShowersHeavy:
                return "Snow Showers (Heavy)"
            case WeatherCode.Thunderstorm:
                return "Thunderstorm"
            case WeatherCode.ThunderstormWithHailSight:
                return "Thunderstorm With Hail (Sight)"
            case WeatherCode.ThunderstormWithHailHeavy:
                return "Thunderstorm With Hail (Heavy)"
            case WeatherCode.Unknown:
                return "Unknown"
            case _:
                return "Unsupported (bug?)"

    @staticmethod
    def code_to_icon(code: WeatherCode):
        match code:
            case WeatherCode.ClearSky | WeatherCode.MainlyClear:
                return ctx.icons.sunny
            case WeatherCode.PartlyCloudy | WeatherCode.Overcast:
                return ctx.icons.partly_cloudy
            case WeatherCode.FogDepositingRime | WeatherCode.FogNormal:
                return ctx.icons.foggy
            case (
                WeatherCode.DrizzleDense
                | WeatherCode.DrizzleLight
                | WeatherCode.DrizzleModerate
                | WeatherCode.FreezingDrizzleLight
                | WeatherCode.FreezingDrizzleDense
                | WeatherCode.RainSlight
                | WeatherCode.RainModerate
                | WeatherCode.RainHeavy
                | WeatherCode.FreezingRainLight
                | WeatherCode.FreezingRainHeavy
                | WeatherCode.RainShowersSlight
                | WeatherCode.RainShowersModerate
                | WeatherCode.RainShowersViolent
            ):
                return ctx.icons.rainy
            case (
                WeatherCode.SnowFallSlight
                | WeatherCode.SnowFallModerate
                | WeatherCode.SnowFallHeavy
                | WeatherCode.SnowGrains
                | WeatherCode.SnowShowersSlight
                | WeatherCode.SnowShowersHeavy
            ):
                return ctx.icons.snowy
            case (
                WeatherCode.Thunderstorm
                | WeatherCode.ThunderstormWithHailSight
                | WeatherCode.ThunderstormWithHailHeavy
            ):
                return ctx.icons.thunderstorm
            case WeatherCode.Unknown:
                return ctx.icons.question_mark
